models/classes/search/ResultSetResponseNormalizer.php

```diff
@@ -91,14 +82,78 @@ public function normalize(SearchQuery $searchQuery, ResultSet $resultSet, string
 
                 $resourcePermissions[$resourceId] = !$hasReadAccess;
 
-                $response['data'][] = $content;
+                $responseData[] = $content;
             }
         }
 
+        return $this->createResponse(
+            $responseData ?? [],
+            $resourcePermissions,
+            $searchQuery,
+            $resultSet,
+            $resultAmount
+        );
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function normalizeSafeClass(SearchQuery $searchQuery, ResultSet $resultSet, string $structure): array
+    {
+        $resultsRaw = $resultSet->getArrayCopy();
+        $resultAmount = count($resultsRaw);
+        $resourcePermissions = [];
+
+        if ($resultAmount > 0) {
+            $accessibleResultsMap = array_flip(
+                $this->getPermissionHelper()
+                    ->filterByPermission(
+                        array_column($resultsRaw, 'id'),
+                        PermissionInterface::RIGHT_READ
+                    )
+            );
+
+            foreach ($resultsRaw as $content) {
+                if (!is_array($content)) {
+                    $this->logError(
+                        sprintf(
+                            'Search content issue detected: expected array, but %s given',
+                            json_encode($content)
+                        )
+                    );
+                    continue;
+                }
+
+                $isAccessible = isset($accessibleResultsMap[$content['id']]);
+                $resourcePermissions[$content['id']] = !$isAccessible;
+                $responseData[] = $content;
+            }
+        }
+
+        return $this->createResponse(
+            $responseData ?? [],
+            $resourcePermissions,
+            $searchQuery,
+            $resultSet,
+            $resultAmount
+        );
+    }
+
+    private function createResponse(
+        array $responseData,
+        array $resourcePermissions,
+        SearchQuery $searchQuery,
+        ResultSet $resultSet,
+        int $resultAmount
+    ): array {
+
+        $response['data'] = $responseData;
         $response['readonly'] = $resourcePermissions;
         $response['success'] = true;
         $response['page'] = empty($response['data']) ? 0 : $searchQuery->getPage();
-        $response['total'] = $totalPages;
+        $response['total'] = is_null($searchQuery->getRows()) || $searchQuery->getRows() === 0
+            ? 1
+            : ceil($resultSet->getTotalCount() / $searchQuery->getRows());
```

Comment:
[Optional] Wdyt about this? 🙂
```suggestion
        $response['total'] = empty($searchQuery->getRows())
            ? 1
            : ceil($resultSet->getTotalCount() / $searchQuery->getRows());
```

Reply:
fixed, thanks! :partying_face: